JIRA-OSDOCS3037: Developed content for persistent storage using AWS EBS (#42174)
openshift/openshift-docs: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc

:_content-type: ASSEMBLY
[id="rosa-persistent-storage-aws-ebs"]
= Persistent storage using AWS Elastic Block Store (EBS)
include::modules/attributes-openshift-dedicated.adoc[]
:context: rosa-persistent-storage-aws-ebs

toc::[]

> {product-title} (ROSA) clusters are pre-built with two storage classes that use AWS Elastic Block Store (EBS) volumes. These storage classes are ready to use and some familiarity with Kubernetes and AWS is assumed.

Reviewer: ```suggestion
{product-title} (ROSA) clusters are prebuilt with two storage classes that use AWS Elastic Block Store (EBS) volumes. These storage classes are ready to use and some familiarity with Kubernetes and AWS is assumed.
```

Author: Made the change.

> Following are the two pre-built storage classes:

Reviewer: ```suggestion
Following are the two prebuilt storage classes:
```
This is a sentence fragment, but you could consider just making it a table title (using a `.Heading`) to avoid having to say too much :)

Author: Made the change.

[options="header"]

|===

| Name | Provisioner

| gp2 | kubernetes.io/aws-ebs

| gp2-csi | ebs.csi.aws.com

|===

[NOTE]
====
The gp2 storage class is set as default; however, you can select either one as the default storage class.

> ====

Reviewer: Consider just adding `(default)` next to the `gp2` line in the table, or possibly just making this a regular sentence in one of the paragraphs.

Author: Moved the text into a regular sentence in a paragraph below the table.

> The Kubernetes persistent volume framework enables the administrators to provision a cluster with persistent storage and gives users a way to request those resources without having any knowledge of the underlying infrastructure. You can dynamically provision the AWS EBS volumes. Persistent volumes are not bound to a single project or namespace; therefore, the volumes can be shared across ROSA clusters. Persistent volume claims are specific to a project or namespace and can be requested by users.

Reviewer: ```suggestion
The Kubernetes persistent volume framework enables administrators to provision a cluster with persistent storage and gives users a way to request those resources without having any knowledge of the underlying infrastructure. You can dynamically provision AWS EBS volumes. Persistent volumes are not bound to a single project or namespace; therefore, the volumes can be shared across ROSA clusters. Persistent volume claims are specific to a project or namespace and can be requested by users.
```

Author: Made the change.

[IMPORTANT]
====

> ROSA defaults to using an in-tree or non-Container Storage Interface (CSI) plug-in to provision AWS EBS storage. In future ROSA versions, volumes provisioned using existing in-tree plug-ins are planned for migration to their equivalent CSI driver. After full migration, the in-tree plug-ins will eventually be removed from the future versions of ROSA.

Reviewer: This suggestion applies if "in-tree" = non-CSI (I think it does)
```suggestion
ROSA defaults to using an in-tree, or non-Container Storage Interface (CSI), plug-in to provision AWS EBS storage. In future ROSA versions, volumes provisioned using existing in-tree plug-ins are planned for migration to their equivalent CSI driver. After full migration, the in-tree plug-ins will eventually be removed from the future versions of ROSA.
```
Can you rephrase that final sentence to avoid "will" (definite) and speak more about "plans"?

Author: Made the edit.

====

[IMPORTANT]
====
High-availability of storage in the infrastructure is left to the underlying storage provider.

> ====

Reviewer: Is it possible to restructure to avoid stacked admonitions? If not, consider including them both in one box as a bulleted list.

Author: Thanks! Added both notes in one box as a bulleted list.

> == Format of persistent volumes

Reviewer: This may be different for ROSA docs, but OCP core would require an [id] for each section. Same applies for `== Capacity of EBS volumes on a node`

Author: Added the [id] for each section, and I'll try using this as basic repo hygiene going forward. Thank you!

> Before a ROSA cluster mounts the volume and passes it to a container, the cluster checks that the volume contains a file system as specified by the *fsType* parameter in the persistent volume definition. If the device is not formatted with the file system, all data from the device is erased and the device is automatically formatted with the given file system.

Reviewer: ```suggestion
Before a ROSA cluster mounts the volume and passes it to a container, the cluster checks that the volume contains a file system as specified by the *fsType* parameter in the persistent volume definition. If the device is not formatted with a file system, all data from the device is erased and the device is automatically formatted with the given file system.
```

Author: Made the edit.

This verification enables you to use unformatted AWS volumes as persistent volumes, as the ROSA cluster formats the AWS volumes before the first use.

== Capacity of EBS volumes on a node
By default, a ROSA cluster supports a maximum of 39 EBS volumes attached to one node. This limit is consistent with the link:https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html#linux-specific-volume-limits[AWS volume limits]. The volume limit depends on the instance type.

[IMPORTANT]
====

> You must use either in-tree or CSI volumes and their respective storage classes, but never both volume types at the same time. The maximum attached EBS volume number is counted separately for in-tree and CSI volumes.

Reviewer: > The maximum attached EBS volume number is counted separately for in-tree and CSI volumes.

What does this mean? you could have up to 39 of each type?

Author: That's right. Made the change to clarify this.

====

> == Creating a persistent volume claim

Reviewer: This should be its own procedure module.

Author: Done

.Prerequisites
Storage must exist in the underlying infrastructure before it can be mounted as a volume in the ROSA cluster.

.Procedure

> . In the OpenShift Cluster Manager (OCM) console, click *Storage → Persistent Volume Claims*.

Reviewer 2: It should be OpenShift cluster console rather than 'OpenShift Cluster Manager (OCM) console'.

Author: Made the change, thanks.

. In the persistent volume claims overview, click *Create Persistent Volume Claim*.
. Define the desired options on the page that appears.

>   .. Select the storage class created previously from the drop-down menu.

Reviewer: ```suggestion
  .. Select the previously created storage class from the drop-down menu.
```

Author: Made the change.

  .. Enter a unique name for the storage claim.

>   .. Select the access mode. This selection determines the read and write access for the newly created storage claim.

Reviewer: ```suggestion
  .. Select the access mode. This selection determines the read and write access for the storage claim.
```

Author: Made the change.

  .. Define the size of the storage claim.
. Click *Create* to create the persistent volume claim and generate a persistent volume.

> == Additional resources

Reviewer: Check this against https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#additional-resources-sections
Needs some tagging, should also be a list not a paragraph.

Author: Thanks, made changes as necessary.
